### PR TITLE
Add Integration Test setup

### PR DESCRIPTION
**Pull request by RageCage64**
Working towards closing #6 

* Add IntegrationTest project
* Include .env in build for integration test project
* Use `Program.cs` as WebApplicationFactory entry point
* Use `Staging` environment for integration tests
* db_test now mounts the correct port on the container
* Adjust env var for test port (test DB will now share name)
* Created TestApiFactory, a base class for future integration tests that will set Staging environment and clear database each tests

### CODE REVIEW

File: LeaderboardBackend/Program.cs [hunk 1]
```diff
@@ -67,39 +73,54 @@
...
+	} else
+	{
+		builder.Services.AddDbContext<T>(
+			opt => {
+				opt.UseNpgsql(
+					GetConnectionString(builder)
+				).UseSnakeCaseNamingConvention();
+			}
+		);
+	}
```
Comment (zysim) [inline LeaderboardBackend/Program.cs:117]: May I know why did you make this refactor? Is there an issue with the previous block?

File: LeaderboardBackend/Program.cs [hunk 2]
```diff
@@ -67,39 +73,54 @@
...
 }
+
+#endregion
+
+#region Accessible Program class
+
+public partial class Program { }
```
Comment (zysim) [inline LeaderboardBackend/Program.cs:124]: Why do we need this class for?

File: LeaderboardBackend/appsettings.Staging.json [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+		"LogLevel": {
+			"Default": "Information",
+			"Microsoft.AspNetCore": "Warning"
+		}
+	},
+	"EnvPath": ".env",
+	"Jwt": {
+		"Key": "coolsecretkeywow",
+		"Issuer": "leaderboards.gg"
+	},
```
Comment (zysim) [inline LeaderboardBackend/appsettings.Staging.json:12]: We'll really need to change this ha

File: docker-compose.yml [hunk 4]
```diff
@@ -18,9 +18,9 @@ services:
         environment:
             POSTGRES_USER: ${POSTGRES_USER}
             POSTGRES_PASSWORD: ${POSTGRES_PASSWORD}
-            POSTGRES_DB: ${POSTGRES_TEST_DB}
+            POSTGRES_DB: ${POSTGRES_DB}
         ports:
-            - ${POSTGRES_TEST_PORT}:5433
+            - ${POSTGRES_TEST_PORT}:5432
```
Comment (zysim) [inline docker-compose.yml:23]: Should we really default to the "standard" port? Wouldn't we run the risk of modifying our local DB unexpectedly? Like yeah I guess it doesn't really matter if it's local.

**Review by zysim — COMMENTED** [review]
Mostly just questions from me here learning about .NET.

File: LeaderboardBackend.Integration/LeaderboardBackend.Integration.csproj [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    <PackageReference Include="Microsoft.EntityFrameworkCore.Design" Version="6.0.1">
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+      <PrivateAssets>all</PrivateAssets>
+    </PackageReference>
+    <PackageReference Include="Microsoft.EntityFrameworkCore.InMemory" Version="6.0.1" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="6.0.1" />
+    <PackageReference Include="Microsoft.VisualStudio.Web.CodeGeneration.Design" Version="6.0.1" />
+    <PackageReference Include="Moq" Version="4.16.1" />
+    <PackageReference Include="Npgsql.EntityFrameworkCore.PostgreSQL" Version="6.0.2" />
+    <PackageReference Include="Swashbuckle.AspNetCore" Version="6.2.3" />
```
[the remark this answers is not in the document]
Comment (RageCage64) [reply LeaderboardBackend.Integration/LeaderboardBackend.Integration.csproj:30]: It won't download the packages twice (your system will recognize that you have them downloaded) but they do need to be separately referenced in each project.

File: LeaderboardBackend/Program.cs (hunk 1 above)
Comment (RageCage64) [reply LeaderboardBackend/Program.cs:117]: I think this was a rebasing accident, I'll change it back (although I'll switch it from one-line if statements).

File: LeaderboardBackend/Program.cs (hunk 2 above)
Comment (RageCage64) [reply LeaderboardBackend/Program.cs:124]: This is necessary to make it a visible internal to the IntegrationTest project. It needs to be able to use this script as an entry-point for the `WebApplicationFactory`. Stolen from https://docs.microsoft.com/en-us/aspnet/core/test/integration-tests?view=aspnetcore-6.0#basic-tests-with-the-default-webapplicationfactory

File: LeaderboardBackend/appsettings.Staging.json (hunk 3 above)
Comment (RageCage64) [reply LeaderboardBackend/appsettings.Staging.json:12]: In dev and staging it doesn't really matter, but we should probably put at least a cryptographically generated key for consistency with eventual production 😄

File: docker-compose.yml (hunk 4 above)
Comment (RageCage64) [reply docker-compose.yml:23]: The order here is:
`<port on your machine>`:`<port on the docker machine>`
The database inside the postgres docker container always runs on 5432, and we want to map that port to the one on the running machine. The way it was before meant we were mapping port 5433 inside the container to something on our machine, on which obviously nothing is running. The 5433 was actually a mistake that I left in back in the Go days and copy pasted when I moved the docker stuff over LOL

File: LeaderboardBackend/Program.cs (hunk 1 above)
Comment (RageCage64) [reply LeaderboardBackend/Program.cs]: Changed it back.

File: LeaderboardBackend.Integration/LeaderboardBackend.Integration.csproj (hunk 5 above)
[the remark this answers is not in the document]
Comment (zysim) [reply LeaderboardBackend.Integration/LeaderboardBackend.Integration.csproj:30]: Wack. Fair enough.

File: LeaderboardBackend/Program.cs (hunk 2 above)
Comment (zysim) [reply LeaderboardBackend/Program.cs:126]: Ah, interesting.

File: docker-compose.yml (hunk 4 above)
Comment (zysim) [reply docker-compose.yml:23]: _Oh yeah_ that's the actual syntax. Totally forgot. As you can probably tell, I was thinking the syntax is `<port>:<default>`.

**Review by zysim — APPROVED** [review]

**Pull request merged**